metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,200 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.Applicative
+import cats.data.{NonEmptyList, NonEmptyVector, OptionT}
+import cats.effect.concurrent.Ref
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.CheckpointingService.LedgerTree
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync: Applicative, N](
+    ledgerTree: Ref[F, LedgerTree],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    OptionT(updateLedger(block)).flatMapF { ledger =>
+      val hashValid = ledger.hash == block.header.postStateHash
+      validateTransactions(block.body, ledger).map {
```

Comment:
Ok, I misunderstood that.

So I could just read the `ledgerTree` ref for the parent state. The problem is that if for some reason we validate the last committed block (after we did `executeBlock` on it) then parent state will be gone and the validation will fail. Do you think that's acceptable? 

Alternatively I  guess I could make `updateLedger` return the parent state as well.

Reply:
Right, you mean that `updateLedger` goes back until it finds a state in memory, then builds up returns the last ledger, while inserting it the interim ledgers into memory; at the same time all of it could be removed by the execution. And because the execution removes all the descendants, this can cause a race condition, and if you just tried the read the parent state from the tree, it might not be there any more. 

No, I don't think this should be a reason to fail validation, since we know that this is just a quirk of the current implementation, just incidental complexity, not inherent to the problem.

In the robot example I had a `projectState(block.parentBlockHash)` construct, which was common to `validateBlock`, `executeBlock` and `createBlock` as well, so the state by parent hash is kind of a recurring theme.

Reply:
> No, I don't think this should be a reason to fail validation, since we know that this is just a quirk

The behaviour already is that the ancestors of the last committed block will fail validation even if they were correctly executed before, which is due to the [height condition](https://github.com/input-output-hk/metronome/blob/1a3f15eb976f58996f5249b52991d8585ac81380/metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala#L100). Because the last committed block is also set in stone, I think it's fine to extend this behaviour also to this block.

If you think however that validating the ancestors should always succeed, then that would be a good reason to fix this. Let me have a look at the robot.

Reply:
In the robot example you query the storage quite freely. For efficiency I wanted to avoid that here as much as possible. Do you think that's a misguided optimisation?

Secondly, I was thinking of a possible attack where someone intentionally sends us old blocks to force us to work on reevaluating the state. Hence the height condition. With proper pruning that's not much of a hinderance, but again that would mean that after some threshold validating already executed blocks would fail. So why not, for simplicity, make that threshold equal to the commit?

Going to wait for your input before attempting a fix.

Reply:
Btw, do you think the name `projectLedger` would be better than `updateLedger`? I kinda like the verb "project"

Reply:
> I kinda like the verb "project" 

Yay, I named something right :partying_face: For a moment I wasn't sure what to say, because "project" doesn't seem like a final thing, but then I remembered you have `saveState` when you want to persist something. So maybe `projectLedger` and `saveLedger` could be an intuitive duo?

> In the robot example you query the storage quite freely. For efficiency I wanted to avoid that here as much as possible. Do you think that's a misguided optimisation?

No, I think it's a good goal. I wanted to keep the robot stuff simple (close to being obviously correct, no separate unit tests), since it's just an example, but always thought that we would spend more time on checkpointing to get the most out of it.

> I was thinking of a possible attack where someone intentionally sends us old blocks to force us to work on reevaluating the state. Hence the height condition.

I agree, the height condition is very useful; there is a similar check carried out [here](https://github.com/input-output-hk/metronome/blob/pm-3353-quorum-certificate-phase-typed/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala#L97-L101).

> If you think however that validating the ancestors should always succeed, then that would be a good reason to fix this.

Sorry I think you misunderstood. What I meant is that we shouldn't fail to validate in this scenario:
```
       B4
      / ^ 
     /  validate
    /
B0* - B1* - B2 - B3 
      ^          ^
      exec       validate
```

Say we have commits for `*` blocks, we are currently executing B1 while validating B3 and B4.
When the execution of B1 is done, the current code will clear everything else from the `ledgerTree`, so a code like this could fail:

```scala
for {
  ledger <- OptionT(updateLedger(b3))
  parentLedger <- OptionT(ledgerTreeRef.get.map(_.get(b2.blockHash)))
  isValid <- validate(b3.body, parentLedger)
} yield isValid
```

The problem is that by the time we try to read `parentLedger` it may be gone _not_ because the ancestor B2 is not on the main chain but simply because executing `b1` (which has a commit) has purged the state we just calculated. That's what I called a "quirk".

What you described was the situation for B4, where the disappearance of B0 state means B4 is no longer valid. In that situation, you are right, it doesn't matter any more if we don't consider this block valid any more.

Reply:
I see, but the thing is b3 will be validated correctly anyway, as noted in the [relevant test](https://github.com/input-output-hk/metronome/blob/eb1df073b664df04e1e5249c5fd017795dd57175/metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/CheckpointingServiceProps.scala#L399) (well, I'm 90% sure, now that you suggest otherwise).

The reason is that if the `ledgerTree` gets cleared during validation:
- after the ancestor ledger has been read: then we'll simply apply the descendant blocks (which we already have) to it. We just won't be able to save the intermediate ledgers in the tree, which is wasteful.
- before the ancestor ledger has been read: then we'll keep retrieving blocks from storage until we reach committed ledger in the `ledgerTree` (in this case we'll even be able to save the intermediate ledgers in the tree; only the retrieval is suboptimal).

So the issue you bring is not really of correctness, but of optimisation, and will be handled by the smarter descendant-aware clearing of the `ledgerTree`. (It's already done, but I'm looking if `KVTree` makes it easier)

Reply:
The point I was making was that in your example `B1` would fail validation, after it has already received a `*`. But I would argue that this is not an issue, as there shouldn't be any reason to process an already committed block in any way. Do you agree?

Reply:
We may be talking past each other now. I'm trying to make a point about this comment of yours: 

> So I could just read the ledgerTree ref for the parent state. The problem is that if for some reason we validate the last committed block (after we did executeBlock on it) then parent state will be gone and the validation will fail. 

The "just read `ledgerTree` ref for the parent state"  part in my snippet was `parentLedger <- OptionT(ledgerTreeRef.get.map(_.get(b2.blockHash)))`. This may return `None` because the execution of an ancestor (B1) of the block being validated (B3) could have removed the parent (B2) state, in which case we can falsely conclude that the block is invalid. 

> but the thing is b3 will be validated correctly anyway, as noted in the relevant test

It is true that this test demonstrates that `executeBlock` and `validateBlock` can be called concurrently safely, however, that's for the current logic of `updateLedger` and `validateBlock`, which is not exactly doing the right thing, as per my comment that opened this discussion: you are projecting the state of B3 in the example, and that's done via `updateLedger` returning the chain of ledgers. But you need the parent state! With that change, if it was implemented in the "just read it after calling updateLedger" way, it could fail in some occasions.

> The point I was making was that in your example B1 would fail validation, after it has already received a *

I didn't think B1 would fail validation as in my example it was already being executed, not validated. I was hypothesising about how B3 could fail validation because of the execution of B1, _if_ the code was changed in certain ways :slightly_smiling_face:

Reply:
Right, I see what you mean. The improved `clearLedgerTree` solves this problem too. The required parent will be available for validation as long as it's no older than the last committed ledger.